Add index commit id to searcher

dnhatn opened the pull request:
This commit adds an id, which is composed of the ids of segment files, to ElasticsearchDirectoryReader. With this id, we can retry search requests on another shard copy if its latest "snapshot" has the same segment files as the failing shard.

dnhatn committed:
Assign id to searcher

dnhatn committed:
stylecheck

dnhatn committed:
fix tests

dnhatn committed:
fix tests

dnhatn committed:
more tests

dnhatn commented on server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java:
```diff
@@ -35,12 +42,48 @@
+        return readerId;
+    }
+
+    public static String getReaderIdFromSegmentInfos(SegmentInfos segmentInfos, BooleanSupplier fullyCommitted) {
```
I am not sure if we should pursue this. But without it, we'd never have the same searcher id between copies except with searchable snapshots.

dnhatn commented:
I am labeling this WIP, although it's ready as I am not sure the approach in the PR. The way we compose the searcher id might be too aggressive. If we need to improve search resiliency for searchable snapshots only, then we can use the commit id instead.

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Engine)

henningandersen reviewed (commented):
I believe there are two goals we could have with this id:

1. The ability to repeat the fetch phase if it fails. Here we need identical doc-ids on the replica.
2. The ability to continue to use a PIT despite failures in the cluster. Here we do not need identical doc-ids.

I wonder if the second part is better served by `maxSeqNo` instead whenever `gcp==maxSeqNo`? Plus primary term for correctness.

It might be two completely separate ids and thus handled in another PR. I only wanted to raise this here to gather consensus that we cannot come up with a single id or failover scheme that can serve both purposes.

dnhatn commented:
@jimczi If we implement https://github.com/elastic/elasticsearch/issues/56828 using a composition of indexUUID, shardId, and sequence number (instead of internal document ID), then PIT can go with Henning's option-2. WDYT?

dnhatn committed:
Merge branch 'master' into add-searcher-id

dnhatn committed:
new direction

dnhatn committed:
Use es commit id

dnhatn committed:
try extract

dnhatn committed:
Merge branch 'master' into add-searcher-id

dnhatn committed:
Simplify

dnhatn commented:
@henningandersen @jimczi 

We (Jim, Francisco, and I) discussed this PR and agreed to add two IDs (i.e., commitID and seqID) to searchers. I implemented the commitID using an external UUID that is generated when an engine flushes (see https://github.com/elastic/elasticsearch/pull/63963/commits/2f410eb4210ff043ed1abd52bf052107ca398d7a). However, I found that using an external UUID isn't better than the id of an index commit. Hence, I re-implemented it using the id of an index commit. Would you please take a look? Thank you!

jimczi reviewed (approved):
The logic looks good to me.

henningandersen reviewed (approved):
LGTM.

Thanks for the reminder, @dnhatn

dnhatn committed:
Merge branch 'master' into add-searcher-id

# Conflicts:
#	server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java

dnhatn committed:
base64

dnhatn committed:
javadocs

dnhatn commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1205,6 +1205,15 @@ public final void close() {
+
+        /**
+         * Returns a commit id associated with this searcher if it's opened from an index commit; otherwise, return null.
+         * Two searcher with the same commit id must have identical content at the Lucene document level.
```
see https://github.com/elastic/elasticsearch/pull/63963/commits/193b631a5c6f48e6a12beb1e0e6b04fee97c5fb9

dnhatn commented:
> Could we also set replicas=1 and then check that the recovered copy has the same commit id?
 
We flush after copying segment files in peer recovery to [associate](https://github.com/elastic/elasticsearch/blob/3b859f56a9589de0aa1f32f35364906847804b0a/server/src/main/java/org/elasticsearch/index/store/Store.java#L1455) a new translog UUID, and this flush changes the commit id.

We can avoid this limitation by having an external commit id, and change it whenever the InternalEngine performs a flush. I implemented this, but I reverted it as it helps only this situation. I am happy to bring it back if you prefer.

henningandersen commented:
> but I reverted it as it helps only this situation

Is the purpose of the change not that we can fail over a search to another replica copy containing the same data? I suppose the approach taken here could work for searchable snapshots, but it would be nice to have an approach that also works for shards that already went through file based peer recovery and did not see any changes since (like most warm indices today). Or is there a good reason that we cannot/should not handle that scenario?

s1monw commented on server/src/main/java/org/elasticsearch/common/lucene/Lucene.java:
```diff
@@ -808,6 +809,10 @@ public void delete() {
         }
     }
 
+    public static String getCommitId(SegmentInfos segmentInfos) {
```
can this get some javadocs?

dnhatn replied:
Sure, added in https://github.com/elastic/elasticsearch/pull/63963/commits/193b631a5c6f48e6a12beb1e0e6b04fee97c5fb9.

s1monw reviewed (approved):
LGTM

dnhatn committed:
Merge branch 'master' into add-searcher-id

dnhatn committed:
add javadocs

dnhatn committed:
Add external commit_id

dnhatn committed:
Revert "Add external commit_id"

This reverts commit 5ec5781c2d56571dd665b38e4ec20f74f36afad5.

dnhatn commented:
> Is the purpose of the change not that we can fail over a search to another replica copy containing the same data? 

Yes, that's the original purpose of this PR.

> I suppose the approach taken here could work for searchable snapshots.

Yes, it works with searchable snapshots.

> It would be nice to have an approach that also works for shards that already went through file based peer recovery and did not see any changes since (like most warm indices today). Or is there a good reason that we cannot/should not handle that scenario?

The problem is that we [flush](https://github.com/elastic/elasticsearch/blob/22b3e87e1bc213eac2245be49937191b7f016729/server/src/main/java/org/elasticsearch/index/store/Store.java#L1456) to create a new index commit to associate with a new local translog on replicas after performing a file-based recovery. This generates a new commit_id, although the content of the index commit doesn't change. To circumvent this, we can have an external commit id that doesn't change in file-based recoveries (please see https://github.com/elastic/elasticsearch/pull/63963/commits/5ec5781c2d56571dd665b38e4ec20f74f36afad5).

henningandersen commented:
Thanks @dnhatn , I think we should go for the external commit id like you did in 5ec5781c2d56571dd665b38e4ec20f74f36afad5.

henningandersen commented:
Chatted with @dnhatn about the main issue of the external commit id: relying on lucene not merging when associating a new translog with the commit. In order to continue this path of development, let us stick to the approach you have here (using the commit id of the segment infos), or revisit your original proposal of relying on individual segment commit-ids.

dnhatn commented:
Thanks everyone for reviewing. I am merging this PR as is and will revisit the previous approach later.
